bots/sg_modify_scope_by_port.py

```diff
@@ -0,0 +1,187 @@
+"""
+
+
+def is_rule_exists_in_sg(sg, rule, direction, scope):
+    if direction.lower() == 'inbound':
+        for perm in sg.ip_permissions:
+            if perm['FromPort'] == rule[PORT_FROM] and perm['ToPort'] == rule[PORT_TO] and \
+                    perm['IpProtocol'] == rule[PROTOCOL].lower():
+                for ip in perm['IpRanges']:
+                    if ip['CidrIp'] == scope:  # found same rule exists
+                        return True;
+        return False
+    if direction.lower() == 'outbound':
+        for perm in sg.ip_permissions_egress:
+            if perm['FromPort'] == rule[PORT_FROM] and perm['ToPort'] == rule[PORT_TO] and \
+                    perm['IpProtocol'] == rule[PROTOCOL].lower():
+                for ip in perm['IpRanges']:
+                    if ip['CidrIp'] == scope:  # found same rule exists
+                        return True;
+        return False
+
+
+"""
+checks for ip validity , fix it to be right
+"""
+
+
+def fix_malformed(rule):
+    ip = re.split('/|\.', str(rule[SCOPE]))  # break ip to blocks
+    rule[SCOPE] = ip[0] + '.' + ip[1] + '.' + ip[2] + '.' + ip[3] + '/' + ip[4]
+    pass
+
+
+"""
+creates & removes the specified rules from a security group 
+"""
+
+
+def update_sg(sg, sg_id, rule, scope, direction, text_output):
+    # make sure that scope is in CIDR notation for example, 203.0.113.0/24
+    fix_malformed(rule)
+
+    if direction == 'inbound':
+        try:
+            sg.revoke_ingress(
+                CidrIp=rule[SCOPE],
+                FromPort=rule[PORT_FROM],
+                ToPort=rule[PORT_TO],
+                GroupId=sg_id,
+                IpProtocol=rule[PROTOCOL].lower()
+            )
+            text_output = text_output + ' rule : ' + rule[SCOPE] + ',' + str(rule[PORT_FROM]) + ',' + str(
+                rule[PORT_TO]) + ',' + \
+                          str(sg_id) + ',' + rule[PROTOCOL].lower() + ' deleted successfully ;'
+
+        except Exception as e:
+            text_output = text_output + f'Error while trying to delete security group. Error: {e}'
+
+        # check if one is existing first ! avoid duplicates and exception
+        found = is_rule_exists_in_sg(sg, rule, direction, scope)
+        if found:
+            return text_output
+        try:
+
+            sg.authorize_ingress(
+                CidrIp=scope,
+                FromPort=rule[PORT_FROM],
+                ToPort=rule[PORT_TO],
+                GroupId=sg_id,
+                IpProtocol=rule[PROTOCOL].lower()
+            )
+            text_output = text_output + ' rule : ' + scope + ',' + str(rule[PORT_FROM]) + ',' + str(
+                rule[PORT_TO]) + ',' + \
+                          str(sg_id) + ',' + rule[PROTOCOL].lower() + ' created successfully ;'
+
+        except Exception as e:
+            text_output = text_output + f'Error while trying to create security group. Error: {e}'
+
+    elif direction == 'outbound':
```

Comment:
code duplication?

Reply:
the names of the functions are different for inbound rules and outbound so there is no way to avoid this duplicate 
for inbound is : .revoke_ingress  & authorize_ingress
for outbound is: revoke_egress & authorize_egress

Reply:
So you have IpPermissions declared the same way twice

Reply:
done